src/config.c

```diff
@@ -687,63 +632,151 @@ void loadServerConfig(char *filename, char config_from_stdin, char *options) {
+                for (j = 0; j < i; j++) {
+                    if (set_configs[j] == config) {
+                        /* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
+                        errstr = "duplicate parameter";
+                        invalid_args = 1;
+                        break;
+                    }
+                }
+                set_configs[i] = config;
+                new_values[i] = c->argv[2+i*2+1]->ptr;
+                break;
+            }
+        }
+        /* Fail if we couldn't find this config */
+        /* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
+        if (!set_configs[i]) {
+            errstr = "unrecognized parameter";
+            invalid_args = 1;
+        }
+    }
+    
+    if (invalid_args) goto badfmt;
+
+    /* Backup old values before setting new ones */
+    for (i = 0; i < config_count; i++)
+        old_values[i] = set_configs[i]->interface.get(set_configs[i]->data);
+
+    /* Set all new values (don't apply yet) */
+    for (i = 0; i < config_count; i++) {
+        int res = performInterfaceSet(set_configs[i], new_values[i], &errstr);
+        if (res < 0) {
+            restoreBackupConfig(set_configs, old_values, i+1, NULL);
+            goto badfmt;
+        } else if (res > 0) {
+            /* A new value was set, if this config has an apply function then store it for execution later */
+            if (set_configs[i]->interface.apply) {
+                /* Check if this apply function is already stored */
+                for (j = 0; apply_fns[j] != NULL && j < i; j++) {
+                    if (apply_fns[j] == set_configs[i]->interface.apply)
+                        break;
+                }
+                /* Apply function not stored, store it */
+                if (j == i)
+                    apply_fns[j] = set_configs[i]->interface.apply;
             }
-            if (config->flags & MULTI_ARG_CONFIG) sdsfreesplitres(argv, argc);
-            addReply(c,shared.ok);
-            return;
         }
     }
 
-    addReplyErrorFormat(c,"Unsupported CONFIG parameter: %s",
-        (char*)c->argv[2]->ptr);
-    return;
+    /* Apply all configs after being set */
+    for (i = 0; i < config_count && apply_fns[i] != NULL; i++) {
+        if (!apply_fns[i](&errstr)) {
+            serverLog(LL_WARNING, "Failed applying new %s configuration, restoring previous settings.", set_configs[i]->name);
+            restoreBackupConfig(set_configs, old_values, config_count, apply_fns);
+            goto badfmt;
+        }
+    }
+    addReply(c,shared.ok);
+    goto end;
 
 badfmt: /* Bad format errors */
     if (errstr) {
-        addReplyErrorFormat(c,"Invalid argument '%s' for CONFIG SET '%s' - %s",
-                (char*)o->ptr,
-                (char*)c->argv[2]->ptr,
-                errstr);
+        addReplyErrorFormat(c,"Invalid arguments - %s", errstr);
```

Comment:
is it an "invalid argument"? it could be any other type of (apply) failure?

Reply:
The assumption is that if we failed it's because the input arguments to `CONFIG SET` were bad. So it's always a bad argument issue and there's the extra `errstr` to help here. For example if you try to `CONFIG SET dir /some_dir_that_doesnt_exist` then I can claim the path argument was invalid and I get this:
`ERR Invalid arguments - No such file or directory`. If you really think this is wrong I'll change it to something like `Config set failed - %s`. Let me know.

Reply:
i think `Config set failed - %s` is better, since i think some failures are due to runtime problem (port already in use or such), so it's not really an invalid config, it's just that it failed.